packages/core-compat-api/src/collectLegacyRoutes.tsx

```diff
@@ -58,58 +62,150 @@ Existing tasks:
 
 */
 
+// Creates a shim extension whose purpose is to build up the tree (anchored at
+// the root page) of paths/routeRefs so that the app can bind them properly.
+function makeRoutingShimExtension(options: {
+  name: string;
+  parentExtensionId: string;
+  routePath?: string;
+  routeRef?: RouteRef;
+}) {
+  const { name, parentExtensionId, routePath, routeRef } = options;
+  return createExtension({
+    kind: 'routing-shim',
+    name,
+    attachTo: { id: parentExtensionId, input: 'childRoutingShims' },
+    inputs: {
+      childRoutingShims: createExtensionInput({
+        routePath: coreExtensionData.routePath.optional(),
+        routeRef: coreExtensionData.routeRef.optional(),
+      }),
+    },
+    output: {
+      routePath: coreExtensionData.routePath.optional(),
+      routeRef: coreExtensionData.routeRef.optional(),
+    },
+    factory: () => ({
+      routePath,
+      routeRef: routeRef ? convertLegacyRouteRef(routeRef) : undefined,
+    }),
+  });
+}
+
+function visitRouteChildren(options: {
+  children: ReactNode;
+  parentExtensionId: string;
+  context: {
+    pluginId: string;
+    extensions: ExtensionDefinition<unknown>[];
+    getUniqueName: () => string;
+    discoverPlugin: (plugin: LegacyBackstagePlugin) => void;
+  };
+}): void {
+  const { children, parentExtensionId, context } = options;
+  const { pluginId, extensions, getUniqueName, discoverPlugin } = context;
+
+  Children.forEach(children, node => {
+    if (!isValidElement(node)) {
+      return;
+    }
+
+    const plugin = getComponentData<LegacyBackstagePlugin>(node, 'core.plugin');
+    const routeRef = getComponentData<RouteRef<AnyRouteRefParams>>(
+      node,
+      'core.mountPoint',
+    );
+    const routePath: string | undefined = node.props?.path;
+
+    if (plugin) {
+      // We just mark the plugin as discovered, but don't change the context
+      discoverPlugin(plugin);
+    }
+
+    let nextParentExtensionId = parentExtensionId;
+    if (routeRef || routePath) {
+      const nextParentExtensionName = getUniqueName();
+      nextParentExtensionId = `routing-shim:${pluginId}/${nextParentExtensionName}`;
+      extensions.push(
+        makeRoutingShimExtension({
+          name: nextParentExtensionName,
+          parentExtensionId,
+          routePath,
+          routeRef,
+        }),
+      );
+    }
+
+    visitRouteChildren({
+      children: node.props.children,
+      parentExtensionId: nextParentExtensionId,
+      context,
+    });
+  });
+}
+
 /** @public */
 export function collectLegacyRoutes(
   flatRoutesElement: JSX.Element,
 ): BackstagePlugin[] {
-  const createdPluginIds = new Map<
+  const pluginExtensions = new Map<
     LegacyBackstagePlugin,
     ExtensionDefinition<unknown>[]
   >();
 
+  const getUniqueName = (() => {
+    let currentIndex = 1;
+    return () => String(currentIndex++);
+  })();
```

Comment:
This one doesn't feel super helpful even compared to the nested indices, but I assume the idea is to iterate on this anyway and try to get the old extension name in there. Perhaps put a // TODO?

Reply:
You'll note in the code that it actually doesn't wastefully add layers unless needed. So the parent can get carrier several steps down in the recursion, the old naming felt like it just made less sense then. At this point the number is just there to make sure we avoid collisions.

Reply:
Oh but based on the original extension name - yeah, sure, added a note